mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -1,471 +1,478 @@
-use core::fmt;
 use std::{
-    fmt::{Debug, Formatter},
+    fmt::Debug,
+    ops::{Deref, DerefMut},
     ptr::NonNull,
 };
 
 use libc::c_void;
+use nix::errno::Errno;
 
 use spdk_sys::{
-    spdk_bdev_free_io,
     spdk_bdev_io,
-    spdk_bdev_io_complete,
-    spdk_bdev_io_get_io_channel,
+    spdk_bdev_readv_blocks,
+    spdk_bdev_unmap_blocks,
+    spdk_bdev_write_zeroes_blocks,
+    spdk_bdev_writev_blocks,
     spdk_io_channel,
 };
 
 use crate::{
     bdev::{
         nexus::{
-            nexus_bdev::{Nexus, NEXUS_PRODUCT_ID},
-            nexus_channel::DrEvent,
-            nexus_fn_table::NexusFnTable,
+            nexus_bdev::NEXUS_PRODUCT_ID,
+            nexus_channel::{DrEvent, NexusChannel, NexusChannelInner},
         },
         nexus_lookup,
         ChildState,
+        Nexus,
         NexusStatus,
         Reason,
     },
-    core::{Bdev, Cores, GenericStatusCode, Mthread, NvmeStatus, Reactors},
+    core::{
+        Bdev,
+        BdevHandle,
+        Bio,
+        Cores,
+        GenericStatusCode,
+        IoStatus,
+        IoType,
+        Mthread,
+        NvmeStatus,
+        Reactors,
+    },
+    ffihelper::FfiResult,
     nexus_uri::bdev_destroy,
 };
 
-/// NioCtx provides context on a per IO basis
-#[derive(Debug, Clone)]
-pub struct NioCtx {
-    /// read consistency
-    pub(crate) in_flight: i8,
-    /// status of the IO
-    pub(crate) status: IoStatus,
-    /// attempts left
-    pub(crate) io_attempts: i32,
+#[allow(unused_macros)]
+macro_rules! offset_of {
+    ($container:ty, $field:ident) => {
+        unsafe { &(*(0usize as *const $container)).$field as *const _ as usize }
+    };
 }
 
-impl NioCtx {
-    #[inline]
-    pub fn dec(&mut self) {
-        self.in_flight -= 1;
-        debug_assert!(self.in_flight >= 0);
-    }
+#[allow(unused_macros)]
+macro_rules! container_of {
```

Comment:
Is this macro used in this module ? I wasn't able to find any occurrences of it.

Reply:
I use it during development to validate that I'm holding pointers to what I think I am =) I can remove it if you like but I found it rather useful.